File: Management/Models/Local/LocalCoursePage.cs
namespace LocalModels;

public record LocalCoursePage : IModuleItem
{
  public required string Name { get; init; }
  public required string Text { get; set; }
  public DateTime DueAt { get; init; }
  public string GetBodyHtml() => Markdig.Markdown.ToHtml(Text);

  public string ToMarkdown()
  {
    var printableDueDate = DueAt.ToString()?.Replace('\u202F', ' ');
    var settingsMarkdown = $"Name: {Name}\n"
      + $"DueDateForOrdering: {printableDueDate}\n"
      + "---\n";
    return settingsMarkdown + Text;
  }
  public static LocalCoursePage ParseMarkdown(string pageMarkdown)
  {
    var rawSettings = pageMarkdown.Split("---")[0];
    var name = MarkdownUtils.ExtractLabelValue(rawSettings, "Name");
    var rawDate = MarkdownUtils.ExtractLabelValue(rawSettings, "DueDateForOrdering");

    DateTime parsedDate = DateTime.TryParse(rawDate, out DateTime parsedDueAt)
      ? parsedDueAt
      : throw new LocalPageMarkdownParseException($"could not parse due date: {rawDate}");


    var text = pageMarkdown.Split("---\n")[1];

    return new LocalCoursePage
    {
      Name = name,
      DueAt = parsedDate,
      Text = text
    };
  }

}


public class LocalPageMarkdownParseException : Exception
{
  public LocalPageMarkdownParseException(string message) : base(message)
  {

  }
}

File: nextjs/src/models/local/page/localCoursePage.ts
import { extractLabelValue } from "../assignment/utils/markdownUtils";
import { IModuleItem } from "../IModuleItem";
import { verifyDateOrThrow } from "../timeUtils";

export interface LocalCoursePage extends IModuleItem {
  name: string;
  text: string;
  dueAt: string;
}

export const localPageMarkdownUtils = {
  toMarkdown: (page: LocalCoursePage) => {
    const printableDueDate = verifyDateOrThrow(page.dueAt, "page DueDateForOrdering")
    const settingsMarkdown = `Name: ${page.name}\nDueDateForOrdering: ${printableDueDate}\n---\n`;
    return settingsMarkdown + page.text;
  },

  parseMarkdown: (pageMarkdown: string) => {
    const rawSettings = pageMarkdown.split("---")[0];
    const name = extractLabelValue(rawSettings, "Name");
    const rawDate = extractLabelValue(rawSettings, "DueDateForOrdering");
    const dueAt = verifyDateOrThrow(rawDate, "page DueDateForOrdering");

    const text = pageMarkdown.split("---\n")[1];

    const page: LocalCoursePage = {
      name,
      dueAt,
      text,
    };
    return page;
  },
};
